build/gengo.bzl

```diff
@@ -0,0 +1,310 @@
+# Copyright 2018 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Bazel macros for using the k8s.io/gengo framework to generate zz_generated.FOO.go files.
+#
+# Usage:
+#   1) Create a rule that builds all the generated files, in //build for example
+#       load("//build:gengo.bzl", "k8s_gengo_all")
+#       k8s_gengo_all(
+#           name="deepcopy-sources",
+#           base="zz_generated.deepcopy.go",
+#           tool="//vendor/k8s.io/code-generator/cmd/deepcopy-gen",
+#           match="+k8s:deepcopy-gen=",
+#           flags="--bounding-dirs k8s.io/kubernetes,k8s.io/api",
+#           packages=[
+#             "k8s.io/kubernetes/every/package",
+#             "k8s.io/kubernetes/that/wants",
+#             "k8s.io/kubernetes/a/zz_genrated.deepcopy.go/file",
+#           ],
+#           deps=[
+#             "vendor/github.com/every/package",
+#             "vendor/whatever.io/used/by/packages",
+#           ],
+#      )
+#   2) For every package in the packages list, add the following:
+#       load("//build:gengo.bzl", "k8s_gengo")
+#       k8s_gengo(name="generate-deepcopy", outs=["zz_generated.deepcopy.go"])
+#   3) Be sure to use the //build:go.bzl wrapped version of go_library
+#       load("@io_bazel_rules_go//go:def.bzl", "go_library")  # NO: generate-deepcopy fails
+#       load("//build:go.bzl", "go_library")  # YES: generate-deepcopy succeeds
+#   4) gazelle will now notice the existence of this file and add it to srcs
+#       go_library(
+#           name="go_default_library",
+#           srcs=[..., "zz_generated.deepcopy.go"],
+#           ...
+#       )
+
+load("@io_kubernetes_build//defs:go.bzl", _go_genrule="go_genrule")
+
+def _generate_prefix(tool, tool_flags, out, match):
+  """Set the variables that change for different generators.
+
+  Args:
+    tool: a gengo tool, something like k8s.io/code-generator/cmd/deepcopy-gen
+    tool_flags: unique flags for the tool, something like --bounding-dirs k8s.io/kubernetes
+    out: the basename of the file, something like zz_generated.deepcopy.go
+    match: search packages for files with this tag, something like +k8s:deepcopy-gen=
+  """
+  return """
+  tool='$(location {tool})'
+  tool_flags='{flags}'
+  out='{out}'
+  match='{match}'
+""".format(tool=tool, tool_flags=tool_flags, out=out, match=match)
+
+def _generate_body():
```

Comment:
use ctx.actions.expand_template then ctx.actions.run:

https://docs.bazel.build/versions/master/skylark/lib/actions.html#expand_template
https://docs.bazel.build/versions/master/skylark/lib/actions.html#run

so the bash can be in a bash file :)

Reply:
Moved `_generate_body()` to .sh file